torch/csrc/jit/codegen/cuda/lower_mma.cpp

```diff
@@ -0,0 +1,609 @@
+#include <torch/csrc/jit/codegen/cuda/ir_all_nodes.h>
+#include <torch/csrc/jit/codegen/cuda/kernel_ir_builder.h>
+#include <torch/csrc/jit/codegen/cuda/lower2device.h>
+#include <torch/csrc/jit/codegen/cuda/lower_mma.h>
+#include <torch/csrc/jit/codegen/cuda/lower_utils.h>
+#include <torch/csrc/jit/codegen/cuda/scheduler/mma_utils.h>
+
+#include <sstream>
+
+namespace torch {
+namespace jit {
+namespace fuser {
+namespace cuda {
+
+namespace {
+
+class LowerMMA {
+ public:
+  static std::vector<kir::Expr*> run(std::vector<kir::Expr*> exprs) {
+    LowerMMA lower_mma;
+    return lower_mma.handle(exprs);
+  }
+
+ private:
+  std::vector<kir::Expr*> handle(const std::vector<kir::Expr*>& exprs) {
+    std::vector<kir::Expr*> processed_exprs;
+    for (auto expr : exprs) {
+      if (auto for_loop = dynamic_cast<kir::ForLoop*>(expr)) {
+        processed_exprs.push_back(handle(for_loop));
+      } else if (auto ite = dynamic_cast<kir::IfThenElse*>(expr)) {
+        processed_exprs.push_back(handle(ite));
+      } else {
+        if (isMmaOprandLoad(expr)) {
+          auto kir_smem_read = lowerMmaOperandSmemRead(expr);
+          processed_exprs.insert(
+              processed_exprs.end(),
+              kir_smem_read.begin(),
+              kir_smem_read.end());
+          continue;
+        } else if (isMmaOprandStore(expr)) {
+          auto kir_smem_write = lowerMmaOperandSmemWrite(expr);
+          processed_exprs.insert(
+              processed_exprs.end(),
+              kir_smem_write.begin(),
+              kir_smem_write.end());
+          continue;
+        } else if (isMMAGlobalWrite(expr)) {
+          auto kir_gmem_write = lowerGlobalWriteC(expr);
+          processed_exprs.insert(
+              processed_exprs.end(),
+              kir_gmem_write.begin(),
+              kir_gmem_write.end());
+          continue;
+        } else if (auto maybe_tv = getMaybeInitializedAccumulator(expr)) {
+          // This change is needed because initialization doesn't have its own
+          //  loop structure so will have mma_loops on the initialization
+          //  loopnest.
+          // So will need to manually adjust to define the initializations on
+          //  the mma_loop part.
+          auto acc_tv = maybe_tv.value();
+          auto& gemm_params = acc_tv->gemmParams();
+          auto uop = expr->as<kir::UnaryOp>();
+          // This will depdend on output layout, assume it's n for now.
+          uop->forceVectorize(gemm_params.instruction_tile_shape.n);
+        } else if (auto alloc = dynamic_cast<kir::Allocate*>(expr)) {
+          if (alloc->memoryType() == MemoryType::Shared) {
+            auto buffer_tv = dynamic_cast<kir::TensorView*>(alloc->buffer());
+            if (buffer_tv) {
+              if (lower_mma_utils::isMMAOperandSmem(buffer_tv->fuserTv())) {
+                // Always align to 128b on smem for gemm operands.
+                alloc->align(16);
```

Comment:
I don't think 128b alignment on smem should have any significant drawbacks. I think it's pretty safe, do you have anything specific in mind here?

Reply:
No, but I just think it could be made a little more robust by not using `16`. When `TF32` becomes supported, the alignment number can be `8` even with 128b alignment.

Reply:
Yes in general it'd make sense but this isn't about data type though. Like the comment above said, at the early stage I'd like to go for always vectorize to 16B/128b shared mem loads and stores to ensure performance, at least the loads. 

This is actually closely related to how we define a `CTA tile`, which is very much up to the schedulers. I will need to build out all the rules for custom written ones in another PR.

Reply:
Sorry, never mind. I guess I just had a stupid misunderstanding.